README.md
# StepperComplete

This is a custom Flutter widget that implements a stepper (step-by-step assistant). It allows users to go through a series of steps, with the ability to move forward and backward between the steps.

## Features

- Supports an unlimited number of steps.
- Provides customizable "Continue" and "Back" buttons.
- Allows customization of the colors of the selected and unselected step circles.
- Supports horizontal scrolling when the number of steps is greater than 4.
- Each step can have its own custom content.

## Usage

To use this widget, you need to provide a list of `Step`s. Each `Step` is a widget that represents a single step in the assistant.

## Contributions

Contributions are welcome! If you find a bug or have a feature request, feel free to open an issue. If you want to contribute with code, feel free to open a pull request.

## License

This project is licensed under the terms of the MIT license.

CHANGELOG.md
## 0.0.1

* initial release.

## 0.0.2

* Added comments to the code to improve readability and maintainability.

## 0.0.3

* Corrected existing comments for accuracy and clarity.

## 0.0.4

* Added an example of how to use the `StepperCompleteBase` widget.

## 0.0.5

* Added title below indicator

## 0.0.6

* Added dynamic padding to the widget and button row

## 0.0.7

* Rename widget name

## 0.0.8

* Indicator title correction

## 0.0.9

* Correcting flutter version and configuring usage platform

## 0.1.0

* Improved horizontal scrolling, making scrolling dynamic

## 0.1.1

* Fix pubspec.yaml

test/stepper_complete_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:stepper_complete/stepper_complete.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//   });
// }


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:stepper_complete/stepper_complete.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  int currentStep = 0;
  final formKeys = List.generate(5, (index) => GlobalKey<FormState>());
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
          appBar: AppBar(
            title: const Text('App de Passos'),
            backgroundColor: Colors.blue,
          ),
          body: StepperComplete(
            currentStep: currentStep,
            steps: List.generate(
              5,
              (index) => Step(
                title: Text('Step ${index + 1} + djsk',
                    style: const TextStyle(fontSize: 50)),
                content: Form(
                  key: formKeys[index],
                  child: SingleChildScrollView(
                    child: Column(
                      children: [
                        TextFormField(
                          validator: (value) {
                            if (value == null || value.isEmpty) {
                              return 'Please enter some text';
                            }
                            return null;
                          },
                        ),
                        TextFormField(
                          validator: (value) {
                            if (value == null || value.isEmpty) {
                              return 'Please enter some text';
                            }
                            return null;
                          },
                        ),
                      ],
                    ),
                  ),
                ),
              ),
            ),
            onStepContinue: () {
              setState(() {
                if (currentStep < 19) {
                  currentStep++;
                }
              });
            },
            onStepBack: () {
              setState(() {
                if (currentStep > 0) {
                  currentStep--;
                }
              });
            },
          )),
    );
  }
}


lib/src/stepper_complete_base.dart
import 'package:flutter/material.dart';

class StepperCompleteBase extends StatefulWidget {
  /// The [int] that is currently displayed.
  final int currentStep;

  /// The [Widget] that contains the buttons.
  final Widget? buttons;

  /// The [List<Step>] of the stepper whose titles and icons are displayed
  /// in a horizontal list.
  final List<Step> steps;

  /// The [String] that is displayed when the user is on the current step.
  final String? titleStepBack;

  /// The [String] that is displayed when the user is on the current step.
  final String? titleStepContinue;

  /// The [Color] of the circle for the current step.
  final Color? colorSelectedCircle;

  /// The [Color] of the circles for the steps that have been completed.
  final Color? colorUnselectedCircle;

  /// The [VoidCallback] that is called when the user taps the continue button.
  final VoidCallback? onStepContinue;

  /// The [VoidCallback] that is called when the user taps the back button.
  final VoidCallback? onStepBack;

  /// The [EdgeInsetsGeometry] around the edges of the [Column].
  final EdgeInsetsGeometry? padding;

  /// The [EdgeInsetsGeometry] around the edges of the [Row] of buttons.
  final EdgeInsetsGeometry? paddingButtons;

  const StepperCompleteBase({
    Key? key,
    required this.currentStep,
    required this.steps,
    this.onStepContinue,
    this.onStepBack,
    this.titleStepBack,
    this.titleStepContinue,
    this.buttons,
    this.colorSelectedCircle = Colors.blue,
    this.colorUnselectedCircle = Colors.grey,
    this.padding,
    this.paddingButtons,
  }) : super(key: key);

  @override
  State<StepperCompleteBase> createState() => _StepperCompleteBaseState();
}

class _StepperCompleteBaseState extends State<StepperCompleteBase> {
  /// The [ScrollController] used to control the scrolling of the steps.
  final ScrollController _scrollController = ScrollController();

  @override
  void didUpdateWidget(covariant StepperCompleteBase oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.currentStep != oldWidget.currentStep) {
      if (widget.steps.length > 4) {
        if (widget.currentStep == 0) {
          _scrollController.jumpTo(0.0);
        } else {
          _scrollController.animateTo(
            widget.currentStep * 60.0,
            duration: const Duration(milliseconds: 500),
            curve: Curves.easeInOut,
          );
        }
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: widget.padding ?? const EdgeInsets.symmetric(horizontal: 20),
      child: Column(
        children: [
          SingleChildScrollView(
            controller: _scrollController,
            scrollDirection: Axis.horizontal,
            child: Row(
              children: widget.steps.map((step) {
                int index = widget.steps.indexOf(step);
                bool isCurrentStep = index == widget.currentStep;

                return Row(
                  children: [
                    Column(
                      children: [
                        CircleAvatar(
                          radius: 20,
                          backgroundColor: isCurrentStep
                              ? widget.colorSelectedCircle
                              : widget.colorUnselectedCircle,
                          child: widget.currentStep < index
                              ? Text(
                                  '${index + 1}',
                                  style: const TextStyle(color: Colors.white),
                                )
                              : (widget.currentStep == index
                                  ? Text(
                                      '${index + 1}',
                                      style:
                                          const TextStyle(color: Colors.white),
                                    )
                                  : const Icon(
                                      Icons.check,
                                      color: Colors.white,
                                    )),
                        ),
                        if (widget.currentStep == index)
                          widget.steps[widget.currentStep].title
                      ],
                    ),
                    if (index != widget.steps.length - 1)
                      Container(
                        height: 1,
                        width: 50,
                        color: Colors.grey,
                      ),
                  ],
                );
              }).toList(),
            ),
          ),
          Expanded(
            child: widget.steps[widget.currentStep].content,
          ),
          Padding(
            padding: widget.paddingButtons ??
                const EdgeInsets.symmetric(vertical: 10),
            child: widget.buttons ??
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    ElevatedButton(
                      onPressed: widget.onStepBack,
                      child: Text(widget.titleStepBack ?? 'Back'),
                    ),
                    ElevatedButton(
                      onPressed: widget.onStepContinue,
                      child: Text(widget.titleStepContinue ?? 'Continue'),
                    ),
                  ],
                ),
          ),
        ],
      ),
    );
  }
}


lib/src/stepper_complete.dart
import 'package:flutter/material.dart';

class StepperComplete extends StatefulWidget {
  /// The [int] that is currently displayed.
  final int currentStep;

  /// The [Widget] that contains the buttons.
  final Widget? buttons;

  /// The [List<Step>] of the stepper whose titles and icons are displayed
  /// in a horizontal list.
  final List<Step> steps;

  /// The [String] that is displayed when the user is on the current step.
  final String? titleStepBack;

  /// The [String] that is displayed when the user is on the current step.
  final String? titleStepContinue;

  /// The [Color] of the circle for the current step.
  final Color? colorSelectedCircle;

  /// The [Color] of the circles for the steps that have been completed.
  final Color? colorUnselectedCircle;

  /// The [VoidCallback] that is called when the user taps the continue button.
  final VoidCallback? onStepContinue;

  /// The [VoidCallback] that is called when the user taps the back button.
  final VoidCallback? onStepBack;

  /// The [EdgeInsetsGeometry] around the edges of the [Column].
  final EdgeInsetsGeometry? padding;

  /// The [EdgeInsetsGeometry] around the edges of the [Row] of buttons.
  final EdgeInsetsGeometry? paddingButtons;

  const StepperComplete({
    Key? key,
    required this.currentStep,
    required this.steps,
    this.onStepContinue,
    this.onStepBack,
    this.titleStepBack,
    this.titleStepContinue,
    this.buttons,
    this.colorSelectedCircle = Colors.blue,
    this.colorUnselectedCircle = Colors.grey,
    this.padding,
    this.paddingButtons,
  }) : super(key: key);

  @override
  State<StepperComplete> createState() => _StepperCompleteState();
}

class _StepperCompleteState extends State<StepperComplete> {
  /// The [ScrollController] used to control the scrolling of the steps.
  final ScrollController _scrollController = ScrollController();

  /// The [List<GlobalKey>] of the steps.
  late List<GlobalKey> _stepKeys;

  @override
  void initState() {
    super.initState();
    // Create a list of keys, one for each step.
    _stepKeys = List<GlobalKey>.generate(
      widget.steps.length,
      (_) => GlobalKey(),
    );
  }

  @override
  void didUpdateWidget(covariant StepperComplete oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.currentStep != oldWidget.currentStep) {
      WidgetsBinding.instance.addPostFrameCallback((_) {
        // if there are more than 4 steps, scroll to the middle of the current step
        if (widget.steps.length > 4) {
          if (widget.currentStep == 0) {
            _scrollController.jumpTo(0.0);
          } else {
            // calculate the offset by adding up the widths of previous steps
            double offset =
                _stepKeys.sublist(0, widget.currentStep).map((GlobalKey key) {
              final RenderBox currentStepRenderBox =
                  key.currentContext!.findRenderObject() as RenderBox;
              return currentStepRenderBox.size.width;
            }).reduce((double value, double element) => value + element);
            // calculate the width of the current step
            final RenderBox currentStepRenderBox = _stepKeys[widget.currentStep]
                .currentContext!
                .findRenderObject() as RenderBox;
            final double stepWidth = currentStepRenderBox.size.width;
            // calculate the width of the screen
            final double screenWidth = MediaQuery.of(context).size.width;
            // scroll to the middle of the current step
            _scrollController.animateTo(
              offset - screenWidth / 2 + stepWidth / 2,
              duration: const Duration(milliseconds: 500),
              curve: Curves.easeInOut,
            );
          }
        }
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: widget.padding ?? const EdgeInsets.symmetric(horizontal: 20),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.spaceAround,
        children: [
          SingleChildScrollView(
            controller: _scrollController,
            scrollDirection: Axis.horizontal,
            child: Row(
              children: widget.steps.map((step) {
                // get the index of the step
                int index = widget.steps.indexOf(step);
                // check if the step is the current step
                bool isCurrentStep = index == widget.currentStep;

                return Row(
                  key: _stepKeys[index],
                  children: [
                    Padding(
                      padding: const EdgeInsets.only(top: 10, bottom: 10),
                      child: Column(
                        children: [
                          CircleAvatar(
                            radius: 20,
                            backgroundColor: isCurrentStep
                                ? widget.colorSelectedCircle
                                : widget.colorUnselectedCircle,
                            child: widget.currentStep < index
                                ? Text(
                                    '${index + 1}',
                                    style: const TextStyle(color: Colors.white),
                                  )
                                : (widget.currentStep == index
                                    ? Text(
                                        '${index + 1}',
                                        style: const TextStyle(
                                            color: Colors.white),
                                      )
                                    : const Icon(
                                        Icons.check,
                                        color: Colors.white,
                                      )),
                          ),
                          if (widget.currentStep == index)
                            widget.steps[widget.currentStep].title
                        ],
                      ),
                    ),
                    if (index != widget.steps.length - 1)
                      Container(
                        margin: const EdgeInsets.symmetric(horizontal: 5),
                        height: 1,
                        width: 50,
                        color: Colors.grey,
                      ),
                  ],
                );
              }).toList(),
            ),
          ),
          Expanded(
            child: widget.steps[widget.currentStep].content,
          ),
          Padding(
            padding: widget.paddingButtons ??
                const EdgeInsets.symmetric(vertical: 10),
            child: widget.buttons ??
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    ElevatedButton(
                      onPressed: widget.onStepBack,
                      child: Text(widget.titleStepBack ?? 'Back'),
                    ),
                    ElevatedButton(
                      onPressed: widget.onStepContinue,
                      child: Text(widget.titleStepContinue ?? 'Continue'),
                    ),
                  ],
                ),
          ),
        ],
      ),
    );
  }
}
